File: fn/src/main/java/org/sqlbuilder/fn/joins/LexUnit_SemType.java
package org.sqlbuilder.fn.joins;

import org.sqlbuilder.common.Insertable;
import org.sqlbuilder.fn.objects.Governor;

import java.util.Comparator;
import java.util.HashSet;
import java.util.Set;

import edu.berkeley.icsi.framenet.SemTypeRefType;

/*
lexunits_semtypes.table=fnlexunits_semtypes
lexunits_semtypes.create=CREATE TABLE IF NOT EXISTS %Fn_lexunits_semtypes.table% ( luid INTEGER NOT NULL,semtypeid INTEGER NOT NULL,PRIMARY KEY (luid,semtypeid) );
lexunits_semtypes.fk1=ALTER TABLE %Fn_lexunits_semtypes.table% ADD CONSTRAINT fk_%Fn_lexunits_semtypes.table%_luid FOREIGN KEY (luid) REFERENCES %Fn_lexunits.table% (luid);
lexunits_semtypes.fk2=ALTER TABLE %Fn_lexunits_semtypes.table% ADD CONSTRAINT fk_%Fn_lexunits_semtypes.table%_semtypeid FOREIGN KEY (semtypeid) REFERENCES %Fn_semtypes.table% (semtypeid);
lexunits_semtypes.no-fk1=ALTER TABLE %Fn_lexunits_semtypes.table% DROP CONSTRAINT fk_%Fn_lexunits_semtypes.table%_luid CASCADE;
lexunits_semtypes.no-fk2=ALTER TABLE %Fn_lexunits_semtypes.table% DROP CONSTRAINT fk_%Fn_lexunits_semtypes.table%_semtypeid CASCADE;
lexunits_semtypes.insert=INSERT INTO %Fn_lexunits_semtypes.table% (luid,semtypeid) VALUES(?,?);
 */
public class LexUnit_SemType extends Pair<Integer, Integer> implements Insertable<LexUnit_SemType>
{
	public static final Comparator<LexUnit_SemType> COMPARATOR = Comparator.comparing(LexUnit_SemType::getFirst).thenComparing(LexUnit_SemType::getSecond);

	public static final Set<LexUnit_SemType> SET = new HashSet<>();

	// C O N S T R U C T O R

	public static LexUnit_SemType make(final int luid, final SemTypeRefType semtype)
	{
		var ut = new LexUnit_SemType(luid, semtype.getID());
		SET.add(ut);
		return ut;
	}

	private LexUnit_SemType(final int luid, final int semtypeid)
	{
		super(luid, semtypeid);
	}

	// I N S E R T

	@Override
	public String dataRow()
	{
		return String.format("%d,%d", first, second);
	}

	// T O S T R I N G

	@Override
	public String toString()
	{
		return String.format("[LU-SEM luid=%s semtypeid=%s]", first, second);
	}
}

File: fn/src/main/java/org/sqlbuilder/fn/joins/LexUnit_SemType.kt
package org.sqlbuilder.fn.joins

import edu.berkeley.icsi.framenet.SemTypeRefType
import org.sqlbuilder.common.Insertable
import java.util.*

data class LexUnit_SemType(
    val luid: Int,
    val semtypeid: Int,
) : Insertable {

    // I N S E R T

    override fun dataRow(): String {
        return "$luid,$semtypeid"
    }

    // I D E N T I T Y

    override fun equals(o: Any?): Boolean {
        if (this === o) {
            return true
        }
        if (o == null || javaClass != o.javaClass) {
            return false
        }
        val that = o as LexUnit_SemType
        return luid == that.luid && semtypeid == that.semtypeid
    }

    override fun hashCode(): Int {
        return Objects.hash(luid, semtypeid)
    }

    // T O S T R I N G

    override fun toString(): String {
        return "[LU-SEM luid=$luid semtypeid=$semtypeid]"
    }

    companion object {

        val COMPARATOR: Comparator<LexUnit_SemType> = Comparator
            .comparing<LexUnit_SemType, Int> { it.luid }
            .thenComparing<Int> { it.semtypeid }

        val SET = HashSet<LexUnit_SemType>()

        fun make(luid: Int, semtype: SemTypeRefType): LexUnit_SemType {
            val ut = LexUnit_SemType(luid, semtype.getID())
            SET.add(ut)
            return ut
        }
    }
}
